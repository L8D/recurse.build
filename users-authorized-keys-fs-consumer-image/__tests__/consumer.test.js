/* globals describe, it, expect */

const emitterify = require('utilise.emitterify')

const {consumer} = require('../consumer')

const makeMemoryTopicPool = () => new Proxy({}, {
  get: (target, key) => {
    if (key in target) {
      return target[key]
    } else {
      return target[key] = emitterify()
    }
  }
})

describe('users-authorized-keys-fs-consumer', () => {
  it('report a complete job history when the user exists', async () => {
    const topics = makeMemoryTopicPool()

    const props = {
      fero: async (name) => topics[name],
      sh: jest.fn((...args) => ({sh: args})),
      fs: {
        writeFile: jest.fn((...args) => ({writeFile: args}))
      }
    }

    const sinks = await consumer(props)
      
    topics.users.emit('change', {
      key: 'foo',
      value: {authorized_keys: 'baz'}
    })

    sinks.create.stop()

    const output = await sinks.create

    expect(output).toMatchSnapshot()
  })

  it('report a complete job history when the user does not exist', async () => {
    const topics = makeMemoryTopicPool()

    const props = {
      fero: async (name) => topics[name],
      sh: jest.fn((...args) => {
        if (/^id /.test(args[0])) {
          throw new Error('user does not exist')
        }

        return {sh: args}
      }),
      fs: {
        writeFile: jest.fn((...args) => ({writeFile: args}))
      }
    }

    const sinks = await consumer(props)
      
    topics.users.emit('change', {
      key: 'foo',
      value: {authorized_keys: 'baz'}
    })

    sinks.create.stop()

    const output = await sinks.create

    expect(output).toMatchSnapshot()
  })
})
