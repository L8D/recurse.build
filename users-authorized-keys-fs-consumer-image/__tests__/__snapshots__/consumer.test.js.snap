// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users-authorized-keys-fs-consumer report a complete job history when the user does not exist 1`] = `
Array [
  Array [
    Object {
      "sh": Array [
        "adduser -D foo",
      ],
    },
    Object {
      "sh": Array [
        "passwd foo -ud",
      ],
    },
  ],
  Array [
    Object {
      "writeFile": Array [
        "/etc/ssh/authorized-keys/foo",
        "baz",
        Object {
          "mode": 493,
        },
      ],
    },
    Object {
      "sh": Array [
        "mkdir -p -m 0700 /home/foo/.ssh",
      ],
    },
    Object {
      "writeFile": Array [
        "/home/foo/.ssh/authorized_keys",
        "baz",
        Object {
          "mode": 384,
        },
      ],
    },
    Object {
      "sh": Array [
        "chown -R foo /home/foo/.ssh",
      ],
    },
  ],
]
`;

exports[`users-authorized-keys-fs-consumer report a complete job history when the user exists 1`] = `
Array [
  true,
  Array [
    Object {
      "writeFile": Array [
        "/etc/ssh/authorized-keys/foo",
        "baz",
        Object {
          "mode": 493,
        },
      ],
    },
    Object {
      "sh": Array [
        "mkdir -p -m 0700 /home/foo/.ssh",
      ],
    },
    Object {
      "writeFile": Array [
        "/home/foo/.ssh/authorized_keys",
        "baz",
        Object {
          "mode": 384,
        },
      ],
    },
    Object {
      "sh": Array [
        "chown -R foo /home/foo/.ssh",
      ],
    },
  ],
]
`;
